### PR TITLE
auth: key on both subscription id and name on msi login

### LINKED ISSUE #5402: Multiple subscriptions and MSI login, setting default
When MSI for a VM is enabled, and allocated to more than one subscription. Only one subscription is used

az login --msi works and lists all the allocated subscriptions, but it appears to be setting the "name" to "MSI@50342" for all subscriptions. Instead of the actual subscription name.
image

Meaning "az account list --output table" only outputs one result. 
image

And "az account set -s [id/name]" don't work unless you use "MSI@50342" or the last subscription ID in the list.
image

### Environment summary
azure-cli (2.0.23)
Azure Server 2016/ Shell Type (cmd.exe)

### PR DESCRIPTION
**Pull request by yugangw-msft**
Fix #5402 

### General Guidelines

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/5424
This is an **experimental preview** for @microsoft users.

File: src/azure-cli-core/HISTORY.rst [hunk 1]
```diff
@@ -2,9 +2,9 @@
 
 Release History
 ===============
-
 2.0.27
 ++++++
+auth: key on both subscription id and name on msi login
```
Comment (troydai) [inline src/azure-cli-core/HISTORY.rst]: add `*`

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 2]
```diff
@@ -275,28 +275,35 @@ def find_subscriptions_in_vm_with_msi(self, msi_port, identity_id=None):
...
             s[_SUBSCRIPTION_NAME] = "{}@{}".format(base_name, msi_port)
         # key-off subscription name to allow accounts with same id(but under different identities)
-        self._set_subscriptions(consolidated, key_name=_SUBSCRIPTION_NAME)
+        self._set_subscriptions(consolidated, secondary_key_name=_SUBSCRIPTION_NAME)
         return deepcopy(consolidated)
 
-    def _set_subscriptions(self, new_subscriptions, merge=True, key_name=_SUBSCRIPTION_ID):
+    def _set_subscriptions(self, new_subscriptions, merge=True, secondary_key_name=None):
+
+        def _get_key_name(x, secondary_key_name):
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/_profile.py]: Find a better argument name than `x`

File: src/azure-cli-core/azure/cli/core/_profile.py [hunk 3]
```diff
@@ -275,28 +275,35 @@ def find_subscriptions_in_vm_with_msi(self, msi_port, identity_id=None):
...
 
-        dic.update((x[key_name], x) for x in new_subscriptions)
+        dic.update((_get_key_name(x, secondary_key_name), x) for x in new_subscriptions)
         subscriptions = list(dic.values())
         if subscriptions:
             if active_one:
                 new_active_one = next(
-                    (x for x in new_subscriptions if x[key_name] == active_subscription_id),
+                    (x for x in new_subscriptions if x[_SUBSCRIPTION_ID] == active_subscription_id and
+                     (active_secondary_key_val is None or x[secondary_key_name] == active_secondary_key_val)),
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/_profile.py]: This express is too long to follow. Break it into readable pieces?

**Pull request merged**